controlplane/kubeadm/controllers/controller.go

```diff
@@ -160,7 +160,12 @@ func (r *KubeadmControlPlaneReconciler) Reconcile(req ctrl.Request) (res ctrl.Re
 
 		// patch and return right away instead of reusing the main defer,
 		// because the main defer may take too much time to get cluster status
-		if err := patchHelper.Patch(ctx, kcp); err != nil {
+		// Patch ObservedGeneration only if the reconciliation completed successfully
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
```

Comment:
It looks like there is at least one case (possibly multiple) where this would cause setting the observed generation prior to actually reconciling the resource fully.

The first instance I was able to find is where we are adding the finalizer to the resource and returning early